trie/stacktrie.go

```diff
@@ -363,6 +364,15 @@ func (st *StackTrie) hash() []byte {
 func (st *StackTrie) Hash() (h common.Hash) {
 	st.val = st.hash()
 	st.nodeType = hashedNode
+	if len(st.val) < 32 {
+		ret := make([]byte, 32)
+		h := newHasher(false)
+		defer returnHasherToPool(h)
+		h.sha.Reset()
+		h.sha.Write(h.tmp)
```

Comment:
Horrifically nasty bug here! This _works_, but it's _damn_ brittle! You're writing `h.tmp` from a `h` just retrieved from the pool. And it just so happens that it contains what we want it to contain,  since we just returned it to the pool and didn't clear it. Should be 
```
		// If the 'st.val' is < 32, it's not the hash, but the preimage
		// (the  rlp-encoding of the node). For the top level node, we need
		// to force the hashing
		ret := make([]byte, 32)
		h := newHasher(false)
		defer returnHasherToPool(h)
		h.sha.Reset()
		h.sha.Write(st.val)
```

```suggestion
		h.sha.Write(st.val)
```

Reply:
oh wow good catch